File: frontend/src/lib/api/emailReview.ts
/**
 * The review queue — a BU's own view of what email automation is holding for them.
 *
 * Session JWT, not the raw Carmen token: these are our endpoints answering our frontend.
 * `/api/v1/carmen/*` (the settings API Carmen's own server calls) is the other thing, and
 * `emailAutomation.ts` is where that lives.
 */

import { apiFetch } from './client'
import { API } from './endpoints'
import type { Page } from './page'

/** Why a parked document might be worth opening. Computed once at park time and stored,
 *  because neither survives a list query: see `_review_flags` in email_ingest_service.py. */
export type ReviewFlag =
  'unbalanced' | 'mapping_guessed' | 'mapping_missing' | 'doc_no_missing' | 'warnings'

/** Which filter chip a row lives under — see `_chip_expr` in routers/credit_card_activity.py,
 *  which is the only definition there is.
 *
 *  `success` and `unposted` are both about a document this BU **paid to have read**: it
 *  posted, or it did not. An attachment nobody was charged for — a filename rule refusing a
 *  signature logo, a sender the BU never allowed — is in neither, because it is not a
 *  document that failed to post, it is a file that was never one.
 *
 *  `today` and `all` are the two that are not about state. `today` selects on *time* and
 *  overlaps all the others, which is why the server leaves it out of `counts.all`. `all`
 *  selects on nothing at all: it is the module's log, and the only view that shows the
 *  never-charged rows. Both are counted, neither carries a dot. */
export type ActivityFilter = 'all' | 'today' | 'review' | 'success' | 'unposted'

/** The strip, left to right: the day, then the order a document moves through, then the log.
 *
 *  The first three are also the fall-through order the page opens on — `today`, then
 *  `review`, then `success`, stopping at the first with anything in it (`useReviewQueue`).
 *  Today reads first because it is the question asked on arrival, and a BU with twelve
 *  documents owed is still never shown a quiet morning: an empty chip hands over rather than
 *  holding. `all` is last because nobody arrives asking for everything — it is where you go
 *  when a document you expected is not in any of the others. */
export const ACTIVITY_FILTERS: ActivityFilter[] = ['today', 'review', 'success', 'unposted', 'all']

export interface ReviewDocument {
  id: string
  /** Where the document came in from. `manual` rows are scans someone ran through the
   *  wizard; they only ever appear once posted, so they are always `status: 'posted'`. */
  source: 'email' | 'manual'
  created_at: string | null
  attachment: string
  /** The raw ledger status, not the tab. `problem` covers two of these and the row has
   *  to say which: a document Carmen refused and one a colleague rejected are the same
   *  errand but not the same story. */
  status: string
  bank_code: string | null
  doc_no: string | null

  // Only while pending_review — `_finish` clears the payload these come from, so on a
  // resolved row they are absent rather than stale. Rendering `0.00` for a posted
  // document would not be a missing value, it would be a wrong one.
  doc_date: string | null
  /** Gross (Σ pay_amt) — what lands on the credit side of the JV, so what a reviewer
   *  is agreeing to. Not the net. */
  total: number
  line_count: number
  flags: ReviewFlag[]
  /** Payment types nothing could map. The review screen renders an empty picker per
   *  entry; the reviewer fills them and the JV becomes postable. */
  unmapped: string[]
  /** GL rules the AI invented during ingest, by config field type. Marked for checking —
   *  `flags` only says one of them was guessed, which is not enough to point at. */
  guessed: string[]

  // Plain ledger columns, kept for ever.
  jv_no: string | null
  /** Why the pipeline stopped. Once only set on a resolved row — but since a refusal that
   *  came *after* a paid-for extraction parks instead of finishing, a `pending_review` row
   *  can carry one too, and then it is the strongest thing the row has to say. */
  reason_code: string | null
  error_message: string | null
  reviewed_by_name: string | null
  reviewed_at: string | null
  /** Who ran the scan, on a `manual` row only — resolved server-side from the session that
   *  did it, not a stored column, so it goes null once that session has been scrubbed. The
   *  row then says it was scanned by hand without naming anybody, which is the honest
   *  answer; a raw user id would not be. */
  posted_by_name: string | null
}

/** One document, opened. `extracted` is an `/extract` response verbatim, which is exactly
 *  what `useOcrExtraction.applyExtractedData` consumes. */
export interface ReviewDocumentDetail extends ReviewDocument {
  extracted: Record<string, unknown>
}

/** Which of the automation page's four states to render, in one call. `blockers` comes
 *  from the settings service unchanged, so this screen and #/email-settings cannot
 *  disagree about whether a BU is set up. */
export interface ReviewStatus {
  enabled: boolean
  auto_post: boolean
  entitled: boolean
  ingest_address: string | null
  blockers: string[]
  // The endpoint also returns per-tab `counts`, deliberately not typed here: the chips are
  // fed by `listActivity`, which is the only source that also knows about manual scans.
}

/** The activity window plus the counts behind the filter chips. Counts span every row, not
 *  the page — same shape reason `NotificationList` carries `unread_count`. */
export interface ActivityPage extends Page<ReviewDocument> {
  counts: Record<string, number>
  /** Same keys as `counts`: of those rows, how many are wrong in some way — a failure of
   *  any kind, a document claimed and never finished, or a JV that posted without its
   *  input-tax record. Not "what you can fix"; see `_attention` in
   *  routers/credit_card_activity.py for why that was the wrong line to draw.
   *
   *  A lifetime figure. It sizes the pile for the screen-reader sentence; it is **not**
   *  what decides whether the dot shows. */
  attention: Record<string, number>
  /** Same keys again: which chips are holding an anomaly nobody in this BU has looked at.
   *  This is what renders the dot. Nothing retries a failure, so a dot drawn from
   *  `attention` alone would be lit for good — `markChipSeen` is what puts it out. */
  unseen: Record<string, boolean>
}

export async function listActivity(
  filter: ActivityFilter,
  limit = 25,
  offset = 0
): Promise<ActivityPage> {
  const res = await apiFetch(
    `${API.creditCard.activity}?filter=${filter}&limit=${limit}&offset=${offset}`
  )
  if (!res.ok) throw new Error(`Activity fetch failed (${res.status})`)
  return res.json() as Promise<ActivityPage>
}

/**
 * Put that chip's dot out — for everyone in the business unit, not just this browser.
 *
 * The server stores the anomaly count it computes for itself; nothing is sent but the chip
 * name. Failure is silently survivable (the dot simply stays on until next time), so the
 * caller does not await this or surface an error for it.
 *
 * Only a status chip has a mark; `today` and `all` are 400ed by the endpoint. Neither is
 * ever passed, because the caller fires this only for a chip whose dot is lit, and the
 * server sends no `unseen` entry for either.
 */
export async function markChipSeen(filter: ActivityFilter): Promise<void> {
  const res = await apiFetch(API.creditCard.activitySeen, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ filter }),
  })
  if (!res.ok) throw new Error(`Could not mark seen (${res.status})`)
}

/**
 * Put a row away. It leaves the Review chip and stays under Not posted.
 *
 * Nothing is destroyed, which is why the caller offers no confirmation and no undo — the
 * row keeps its whole story, it just stops being work. The server refuses a row that still
 * has a payload: a document waiting for review is Rejected, which records who and why.
 */
export async function dismissRow(id: string): Promise<void> {
  const res = await apiFetch(API.creditCard.activityDismiss(id), { method: 'POST' })
  if (!res.ok) throw new Error(`Could not dismiss (${res.status})`)
}

export async function getPending(id: string): Promise<ReviewDocumentDetail> {
  const res = await apiFetch(API.emailReview.document(id))
  if (!res.ok) throw new Error(`Document fetch failed (${res.status})`)
  return res.json() as Promise<ReviewDocumentDetail>
}

export async function getReviewStatus(): Promise<ReviewStatus> {
  const res = await apiFetch(API.emailReview.status)
  if (!res.ok) throw new Error(`Review status fetch failed (${res.status})`)
  return res.json() as Promise<ReviewStatus>
}

export interface ApproveResult {
  jv_no: string
  /** Set when the JV posted but the input-tax record did not. The document is still
   *  posted and gone from the queue — the VAT is a separate errand, not a failure. */
  tax_note: string | null
}

/**
 * What the reviewer corrected on the input-tax record, as the server names it
 * (`InputTaxOverrides`). `undefined` means "not touched", which is what lets a derived
 * value show through rather than being replaced by an empty string on first render.
 *
 * Only the three things the machine can get wrong and a human can see. The amounts are
 * the JV's and are edited on its table; a record that disagreed with the journal it is
 * filed beside is the one outcome worse than no record. The tax period is likewise absent:
 * it follows the document date, and that is where a misread one is corrected.
 */
export interface ItxOverrides {
  vendor_name?: string
  tax_id?: string
  profile_code?: string
}

export async function approveDocument(
  id: string,
  body: {
    extracted: Record<string, unknown>
    rows: unknown[]
    post_input_tax: boolean
    /** Absent = derive the record, the way the unattended path does. */
    input_tax?: ItxOverrides
  }
): Promise<ApproveResult> {
  const res = await apiFetch(API.emailReview.approve(id), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    // 400 carries Carmen's own rejection message, 409 means someone else got there
    // first. Both are things the reviewer needs to read, so surface the detail.
    const detail = await res
      .json()
      .then(d => (d as { detail?: string }).detail)
      .catch(() => null)
    const err = new Error(detail || `Approve failed (${res.status})`)
    ;(err as Error & { status?: number }).status = res.status
    throw err
  }
  return res.json() as Promise<ApproveResult>
}

export async function rejectDocument(id: string, reason?: string): Promise<void> {
  const res = await apiFetch(API.emailReview.reject(id), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ reason: reason || null }),
  })
  if (!res.ok) throw new Error(`Reject failed (${res.status})`)
}

/** Turn review off, or back on. Its own endpoint: the settings save is a full replace,
 *  so flipping this through it would rewrite rules and passwords as a side effect. */
export async function setAutoPost(on: boolean): Promise<boolean> {
  const res = await apiFetch(API.emailReview.autoPost, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ auto_post: on }),
  })
  if (!res.ok) throw new Error(`Could not save (${res.status})`)
  return ((await res.json()) as { auto_post: boolean }).auto_post
}

File: backend/app/routers/email_review.py
"""The review queue — a BU's own view of what email automation is holding for them.

Separate from `email_automation.py` on purpose. That router answers **Carmen's server**
calling us about settings, so it authenticates by replaying the customer's raw Carmen
token and proving it against their own Carmen (`_caller` / `_resolve`). This one answers
**our own frontend**, where the user already holds a session JWT, so it uses
`get_current_session` like every other tenant-facing route in the app.

What lives here is only what a human needs to act on a parked document. Everything a
machine decided is already on the ledger row and visible at `#/admin/email`.

Design: docs/email-automation/07-human-in-the-loop.md
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import SessionInfo, get_current_session
from app.database import get_db
from app.exceptions import NotFoundError
from app.models.email_automation import EmailDocument
from app.models.identity import Tenant
from app.models.schemas import ExtractedCreditCardData
from app.models.schemas.common import Page
from app.models.schemas.email_automation import (
    ApproveIn,
    ApproveResult,
    RejectIn,
    ReviewDocument,
    ReviewDocumentDetail,
    ReviewStatus,
)
from app.services import email_ingest_service as ingest
from app.services import email_settings_service as es
from app.services.cc_jv import num, r2
from app.utils.pagination import paginate

router = APIRouter(prefix="/api/v1/email", tags=["Email Review"])

PENDING = "pending_review"

# Tab → the ledger statuses under it. Grouped by what the answer means to the person
# looking, not by which code path wrote it:
#
#   review   what needs you now — the only tab with anything to do
#   posted   what went through, which is the evidence the automation is working and the
#            reason "all caught up" is not indistinguishable from "nothing ever arrived"
#   problem  it did not reach Carmen and somebody should know. `failed` (a gate or Carmen
#            said no) and `rejected` (a human said no) differ in who decided, which the
#            row shows — but not in what is now owed, which is the same errand either way.
#   skipped  the BU's own filename and sender rules said "not this file". Free, never
#            charged, and mostly noise — its own tab so it is findable without being in
#            the way of the three that matter.
TABS: dict[str, tuple[str, ...]] = {
    "review": (PENDING,),
    "posted": ("posted",),
    "problem": ("failed", "rejected"),
    "skipped": ("skipped", "received"),
}


def _summarise(row: EmailDocument) -> dict:
    """The parts of a queue row that come out of the stored payload rather than a column.

    A row whose payload has gone (a race with someone else's approve, or a status that
    moved underneath us) still renders — with zeroes, not a 500. The queue's job is to
    show the reviewer what is waiting, and one unreadable row must not blank the page.
    """
    payload = row.review_payload or {}
    extracted = payload.get("extracted") or {}
    details = extracted.get("details") or []
    return {
        "doc_date": extracted.get("doc_date"),
        # Gross, which is what lands on the credit side of the JV.
        "total": r2(sum(num(d.get("pay_amt")) for d in details)),
        "line_count": len(details),
        "flags": list(payload.get("flags") or []),
        # Which payment types nothing could map. The review screen turns these into empty
        # pickers; it cannot re-derive them, because the config it would diff against has
        # moved on since the document parked.
        "unmapped": list(payload.get("unmapped") or []),
        "guessed": list(payload.get("guessed") or []),
    }


def to_review_row(row: EmailDocument) -> ReviewDocument:
    """Public: `credit_card_activity.py` lists these rows beside manual scans and must
    build them the same way, or the two screens disagree about one document."""
    return ReviewDocument(
        id=str(row.id),
        created_at=row.created_at,
        attachment=row.attachment,
        status=row.status,
        bank_code=row.bank_code,
        doc_no=row.doc_no,
        jv_no=row.jv_no,
        reason_code=row.reason_code,
        error_message=row.error_message,
        reviewed_by_name=row.reviewed_by_name,
        reviewed_at=row.reviewed_at,
        **_summarise(row),
    )


@router.get("/documents", response_model=Page[ReviewDocument])
async def list_documents(
    tab: str = Query("review", description="review | posted | problem | skipped"),
    limit: int = Query(25, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    session: SessionInfo = Depends(get_current_session),
):
    """One tab of this BU's mail, newest first.

    `review` is the default because it is the only tab with anything to do; the rest are
    there so a customer can answer "did that one go through?" without asking us. An
    unknown tab falls back to `review` rather than 400ing: the query string is a UI
    detail, and a stale bookmark should land on the useful tab, not on an error.
    """
    statuses = TABS.get(tab, TABS["review"])
    stmt = (
        select(EmailDocument)
        .where(
            EmailDocument.tenant_id == uuid.UUID(str(session.tenant_id)),
            EmailDocument.status.in_(statuses),
        )
        .order_by(EmailDocument.created_at.desc())
    )
    rows, total = await paginate(db, stmt, limit, offset)
    return Page[ReviewDocument](
        total=total, limit=limit, offset=offset, data=[to_review_row(r) for r in rows]
    )


@router.get("/documents/{document_id}", response_model=ReviewDocumentDetail)
async def get_pending(
    document_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    session: SessionInfo = Depends(get_current_session),
):
    """One parked document, with the payload the review screen edits.

    The tenant filter is in the WHERE clause, not an `if` after the fetch: this is the
    only endpoint that hands back extracted line items, and "the row exists but is not
    yours" must be indistinguishable from "no such row".
    """
    row = await db.scalar(
        select(EmailDocument).where(
            EmailDocument.id == document_id,
            EmailDocument.tenant_id == uuid.UUID(str(session.tenant_id)),
            EmailDocument.status == PENDING,
        )
    )
    if row is None:
        raise NotFoundError("This document is not waiting for review")
    payload = row.review_payload or {}
    return ReviewDocumentDetail(
        **to_review_row(row).model_dump(),
        extracted=payload.get("extracted") or {},
        # Read here rather than in `_summarise`: the list endpoint and
        # `credit_card_activity.py` share that helper, and the dept/acc codes are only
        # ever wanted by the screen that lets someone edit them.
        suggested=payload.get("suggested") or {},
    )


@router.get("/status", response_model=ReviewStatus)
async def review_status(
    db: AsyncSession = Depends(get_db),
    session: SessionInfo = Depends(get_current_session),
):
    """Which of the automation page's four states to render, in one call.

    `blockers` is reused verbatim from the settings service rather than re-derived here,
    so the queue screen and the settings screen can never disagree about whether a BU is
    set up.
    """
    tenant = await db.get(Tenant, uuid.UUID(str(session.tenant_id)))
    if tenant is None:
        raise NotFoundError("Unknown business unit")
    row = await es.get_settings(db, tenant)
    body = await es.build_settings_response(db, tenant, row)
    # One grouped count for every tab, in one query rather than four.
    per_status = {
        r.status: r.n
        for r in (
            await db.execute(
                select(EmailDocument.status, func.count().label("n"))
                .where(EmailDocument.tenant_id == tenant.id)
                .group_by(EmailDocument.status)
            )
        ).all()
    }
    counts = {tab: sum(per_status.get(st, 0) for st in statuses) for tab, statuses in TABS.items()}
    return ReviewStatus(
        enabled=bool(body.get("enabled")),
        auto_post=bool(body.get("auto_post")),
        entitled=bool(body.get("entitled")),
        ingest_address=body.get("ingest_address"),
        blockers=list((body.get("status") or {}).get("blockers") or []),
        counts=counts,
    )


@router.post("/documents/{document_id}/approve", response_model=ApproveResult)
async def approve(
    document_id: uuid.UUID,
    body: ApproveIn,
    session: SessionInfo = Depends(get_current_session),
):
    """Post the document the reviewer just checked, under the BU's own credential.

    Synchronous on purpose. Carmen rejects JVs for reasons only a human can fix (a closed
    period, a dept code it does not know), and telling them ten minutes later in a bell
    notification wastes the fact that they are sitting right here. The cost is one slow
    request, which `post_gljv` already is in the wizard.

    No `db` dependency: the service opens its own short sessions around the lock, the
    Carmen call and the ledger write, so a slow Carmen never holds a pooled connection
    open. The pool is 10 for the whole application (Supavisor caps the project at 15).
    """
    result = await ingest.approve_document(
        document_id,
        tenant_id=str(session.tenant_id),
        reviewer=session.carmen_user_id,
        reviewer_name=session.username,
        extracted=ExtractedCreditCardData.model_validate(body.extracted),
        rows=body.rows,
        post_input_tax_record=body.post_input_tax,
        input_tax=body.input_tax,
    )
    return ApproveResult(**result)


@router.post("/documents/{document_id}/reject", status_code=204)
async def reject(
    document_id: uuid.UUID,
    body: RejectIn,
    session: SessionInfo = Depends(get_current_session),
):
    """Terminal, and not a refund — the vision call ran, and that is what the credit
    paid for. The reason is optional free text: a mandatory one gets typed as "x" by day
    three, and an optional one that reaches `#/admin/email` is how we learn what the
    extractor keeps getting wrong."""
    await ingest.reject_document(
        document_id,
        tenant_id=str(session.tenant_id),
        reviewer=session.carmen_user_id,
        reviewer_name=session.username,
        reason=body.reason,
    )
